### DISCUSSION TITLE
Add custom logo

### DISCUSSION THREAD

**Discussion by kylemclaren** [scalar/scalar#914] [Ideas]
Touched on it here: https://github.com/scalar/scalar/discussions/908#discussioncomment-8319698

Would love the ability to specify a `logoUrl`, passed in as a configuration option.

Bonus points for dark and light mode options (perhaps `dark.logoUrl` and `light.logoUrl`).

See Stripe's API [docs](https://stripe.com/docs/api) for inspo:

<img width="344" alt="Screenshot 2024-02-01 at 15 34 56" src="https://github.com/scalar/scalar/assets/3727384/8f322d2e-f05a-471d-b344-4e16e06a2ce4">

Come to think of it, a dark mode toggle like theirs would be neat too.

**Comment by marclave** [comment]
hey @kylemclaren , this is a great idea! We support this on https://docs.scalar.com so we should definitely support it on https://github.com/scalar/scalar.

Also we love fly.io! @cameronrohani made a theme let me know if you're interested in it ✨ (we still need to do a header that you mentioned in the other discussion but this is hosted on scalar.com)
![image](https://github.com/scalar/scalar/assets/6176314/9c1a194e-3d24-49fd-988b-224cf17ed2e5)

https://flytheme.apidocumentation.com/

**Reply by kylemclaren** [reply]
@marclave wild, love the theme @cameronrohani 🤤 

We are using the Kepler theme here: https://docs.machines.dev/

(Very early stages of configuration)

**Reply by cameronrohani** [reply]
Attached is the theme in case you want to use :) 

https://gist.github.com/cameronrohani/fb97c2b085d1b6d518d918b797ca8041

**Reply by kylemclaren** [reply]
Awesome stuff @cameronrohani 

Much appreciated

**Comment by cpbr-dev** [comment]
Hey @marclave, this has been asked a while ago, are there any plans to implement this currently ?

**Comment by DavidTParks** [comment]
Any update here?

**Comment by AndreyMay** [comment]
+1 to this

**Comment by marclave** [comment]
@kylemclaren @DavidTParks @cpbr-dev @DavidTParks 

SO sorry on the delay here, we need to add to the [docs](https://guides.scalar.com)

but you can accomplish a custom header like so 😎 

![Uploading image.png…]()


https://codepen.io/scalarorg/pen/VwOXqam

we are still exploring passing a logo in as a prop! @cameronrohani we should talk about this monday ✨

**Comment by kylemclaren** [comment]
Nice @marclave, thanks. With props, bonus points for light/dark mode options!

**Comment by bdewey-foreflight** [comment]
2 months since the last update. Anything here? What would it take to contribute to this change? Is this feature on the roadmap for 2024?

Or are there docs as to how I can integrate the work around above from @marclave into Hono? e.g. for apiReference() I didn't see a prop for customHTML or customHeader. I see customCss prop but how do I pass the html?

thanks!

**Reply by marclave** [reply]
hey @bdewey-foreflight sorry for the delay!

i think in the short term we should add a config to pass in custom HTML to the integrations rendering https://github.com/scalar/scalar/blob/main/packages/hono-api-reference/src/honoApiReference.ts#L149

@hanspagel thoughts on adding support to pass in a custom header like how we do here https://codepen.io/scalarorg/pen/VwOXqam but into the Hono package to start?

**Reply by slav4ik51493** [reply]
@marclave Is it possible to customize header/logo on AspNetCore? 
As I understand, it's not available to specify logo, but maybe we have some option with customizing header?

Tagging @xC0dex also, as I understand you're working on Asp Net library and may help here. 
BTW, thanks for asp net core port!

**Reply by xC0dex** [reply]
Hi @slav4ik51493, would a solution like provided in the codepen work for you? I could think about adding the option to add some custom HTML elements/CSS snippets to the generated HTML content.

**Reply by ismaelrumzan** [reply]
Really loving how easy it is to set up scalar so far. Only blocker right now is being able to include a custom header as props as we're using next.js app router. I don't think this workaroud https://codepen.io/scalarorg/pen/VwOXqam would work in nextjs as a route

**Reply by slav4ik51493** [reply]
Hi @xC0dex, sorry for the long break. I think it would be great, if you could do custom HTML/CSS snippets for C#.

**Reply by xC0dex** [reply]
@slav4ik51493, I added this option in [this PR](https://github.com/scalar/scalar/pull/4215). I'm not sure if it will stay like this and when it will be merged 👀.

**Reply by fritz-fritz** [reply]
> 2 months since the last update. Anything here? What would it take to contribute to this change? Is this feature on the roadmap for 2024?
> 
> Or are there docs as to how I can integrate the work around above from @marclave into Hono? e.g. for apiReference() I didn't see a prop for customHTML or customHeader. I see customCss prop but how do I pass the html?
> 
> thanks!

Hey was looking into this a bit myself recently. Thought I'd share the direction I'm headed. I just switched over to using `@scalar/hono-api-reference` with my Hono project. I haven't yet added a logo or custom css (kinda digging the dark mode default) but I did fix up the missing opengraph tags! They weren't working as the embeded script loaded them too late.

I intend to leverage this same method to add a nav header with a logo. Essentially, the apiReference() is like other middleware and so you can modify it's response:

```js
api.get('/docs', async (c : any, next) => {
    const scalar = await apiReference({
      pageTitle: c.env.api_title,
      spec: {
        url: `${api_v1}openapi.json`,
      },
      hideDownloadButton: true,
      darkMode: true,
      metaData: {
        title: c.env.api_title,
        description: pkg.description,
        ogTitle: c.env.api_title,
        ogDescription: pkg.description,
        ogImage: c.env.discord_image,
        ogUrl: c.req.host,
        ogType: 'website',
        twitterTitle: c.env.api_title,
        twitterDescription: pkg.description,
        twitterCard: c.env.discord_image,
        twitterSite: c.req.host,
        twitterImage: c.env.discord_image,
      },
      defaultOpenAllTags: true,
      hideClientButton: true,
    })(c, next);
    if ( scalar instanceof Response ) {
      const scalar_html = await scalar.text();
      const modified_html = scalar_html.replace(
        '<meta charset="utf-8" />',
        `<meta charset="utf-8" />
        <meta name="description" content="${pkg.description}">
        <meta property="og:title" content="${c.env.api_title}">
        <meta property="og:description" content="${pkg.description}">
        <meta property="og:image" content="${c.env.discord_image}">
        <meta property="og:url" content="${c.req.host}">
        <meta property="og:type" content="website">
        <meta name="twitter:title" content="${c.env.api_title}">
        <meta name="twitter:description" content="${pkg.description}">
        <meta name="twitter:card" content="${c.env.discord_image}">
        <meta name="twitter:site" content="${c.req.host}">
        <meta name="twitter:image" content="${c.env.discord_image}">
        `
      );
      return new Response(modified_html, { status: scalar.status, headers: scalar.headers });
    } else {
      return scalar;
    }
});
```

**Comment by Trikooo** [comment]
+1

**Comment by sonicmouse** [comment]
I guess this died?

Say what you will about swaggerUI, but at least it was easy to style it.

**Comment by monolithed** [comment]
There're no restrictions on adding a logo or customizing it. I was able to implement this in just a few lines of code (with no luck for mobile view). The real question for the devs is whether it’s worth building a toolbar for this.

**Comment by mhosman** [comment]
It would be great to have this feature.

**Comment by jojech** [comment]
Seconding what people have said here earlier. Especially when using scalar within a tool like Hono.
